### DISCUSSION TITLE
Support Alpine?

### DISCUSSION THREAD

**Discussion by maresb** [mamba-org/micromamba-docker#235] [Show and tell]
I recently did some interesting stuff with the conda-forge [automerge bot](https://github.com/conda-forge/automerge-action/blob/main/Dockerfile).

I was surprised that the following actually worked. (I don't like some of the intermediate stuff, so here's the essential bit):

```Dockerfile
FROM mambaorg/micromamba AS build
RUN micromamba install ... && micromamba clean --all --force-pkgs-dirs --yes

FROM frolvlad/alpine-glibc
ENV MAMBA_ROOT_PREFIX="/opt/conda"
COPY --from=build /opt/conda "${MAMBA_ROOT_PREFIX}"
COPY --from=build-env /bin/micromamba /bin/micromamba  # optional
ENV PATH="$MAMBA_ROOT_PREFIX/bin:$PATH"
```

This is slightly screwy since `bash` isn't prepackaged with Alpine, so you can't "properly" activate with `sh` (unless one also runs `apk add --no-cache bash`). But in practice it works great.

I don't really know anything about glibc or this base container (it was already being used by the automerge bot, which needs to be as small as possible since it's pulled by GH Actions each time). But it seems to me like an interesting use case.

I find two ingredients here interesting:

1. Using frolvlad/alpine-glibc as a base image
2. Using a separate build stage

Might it make sense to consider building images from frolvlad/alpine-glibc?

**Comment by wholtz** [comment]
Interesting.  I'll look into adding `frolvlad/alpine-glibc` (with `bash` installed) as a base image. 

Using a separate build image and copying out the conda environment makes sense if are doing something like building an image for a production micro service, where you will run one program and nothing else. You loose the activation scripts and some flexibility in exchange for a smaller image with less attack surface. Users are of course welcome to do this (and there are some existing tools that build conda based images like this), but I've generally assumed `mambaorg/micromamba` is targeting a use case where a usable shell and the ability to modify environments still exists. I don't see an obvious place where we can benefit internally (within the micromamba-docker project) from separate build stages, but please let me know if you see opportunities.

**Reply by maresb** [reply]
> Interesting. I'll look into adding frolvlad/alpine-glibc (with bash installed) as a base image.

:rocket: 

> and there are some existing tools that build conda based images like this

I wonder if we should mention this, and state the use case more clearly. (OTOH the README is already getting quite big.)

One potential use case could be efficiently creating an image for non-Python software, and the simplicity of being able to install directly into the base image. For instance, the above recipe lets me easily create a GP/PARI image by installing `pari` from conda-forge.

**Comment by wholtz** [comment]
I opened issue #236 to track adding an alpine base-image. In the issue I referenced this discussion and I'm surprised that an automatic link to the issue did not get generated within this discussion.

**Comment by wholtz** [comment]
Alpine base image was added 3 Feb 2023.
